### PR TITLE
Use ExecutableReplicaSet uniformly for running projects

### PR DESCRIPTION
**Pull request by karolz-ms**
This is an improved version of https://github.com/dotnet/aspire/pull/946, including necessary changes to the application builder, dashboard, OTel, and DAPR support

### CODE REVIEW

File: src/Aspire.Hosting.Dapr/DaprOptions.cs [hunk 1]
```diff
@@ -11,13 +11,13 @@ public sealed record DaprOptions
     /// <summary>
     /// Gets or sets the path to the Dapr CLI.
     /// </summary>
-    public string? DaprPath { get; init; }
+    public string? DaprPath { get; set; }
```
Comment (karolz-ms) [inline src/Aspire.Hosting.Dapr/DaprOptions.cs:14]: If it is "`init`", it is impossible to set these options via the new `AddDapr()` callback-based interface.

File: src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public static void AddOtlpEnvironment(IResource resource, IConfiguration configu
 
             // Set the service name and instance id to the resource name and UID. Values are injected by DCP.
             context.EnvironmentVariables["OTEL_RESOURCE_ATTRIBUTES"] = "service.instance.id={{- .UID -}}";
-            context.EnvironmentVariables["OTEL_SERVICE_NAME"] = "{{- .Name -}}";
+            context.EnvironmentVariables["OTEL_SERVICE_NAME"] = "{{- index .Annotations \"otel-service-name\" -}}";
```
Comment (davidfowl) [inline src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs:41]: Why this change?

Comment (karolz-ms) [reply src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs:41]: The reason is, with this change everything is a replica, so the name of the resource has a replica id suffix like `basketservice-i34qqu8`. This does not look that great as an OTel service name.

So I changed the service name to be just a replica set name (e.g. `basketservice`), and to differentiate between different service instances/replicas without resorting to service ID (which is a GUID), I augment it in the UI with the process ID.

File: src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs [hunk 3]
```diff
@@ -97,85 +94,68 @@ public async Task BeforeStartAsync(DistributedApplicationModel appModel, Cancell
...
-                            foreach (var endPoint in endPoints)
-                            {
-                                if (ports.TryGetValue(endPoint.Name, out var value) && value.PortEnvVar is not null)
-                                {
-                                    env.TryAdd(value.PortEnvVar, endPoint.Port.ToString(CultureInfo.InvariantCulture));
-                                }
-                            }
-                        }
+                        env.TryAdd("DAPR_GRPC_PORT", $"{{{{- portFor \"{daprSideCarResourceName}_grpc\" -}}}}");
+                        env.TryAdd("DAPR_HTTP_PORT", $"{{{{- portFor \"{daprSideCarResourceName}_http\" -}}}}");
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/Aspire.Hosting.Dapr/DaprDistributedApplicationLifecycleHook.cs:149]: Agreed

File: src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs [hunk 4]
```diff
@@ -40,6 +43,17 @@ public OtlpApplication(Resource resource, IReadOnlyDictionary<string, OtlpApplic
...
+                case PROCESS_ID:
+                    if (attribute.Value.ValueCase == AnyValue.ValueOneofCase.IntValue)
+                    {
+                        ProcessId = attribute.Value.IntValue;
+                        break;
+                    }
+                    else
+                    {
+                        goto default;
+                    }
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs]: Great suggestion. Fixed.

File: samples/eShopLite/ServiceDefaults/Extensions.cs [hunk 5]
```diff
@@ -56,6 +57,12 @@ public static IHostApplicationBuilder ConfigureOpenTelemetry(this IHostApplicati
                 tracing.AddAspNetCoreInstrumentation()
                        .AddGrpcClientInstrumentation()
                        .AddHttpClientInstrumentation();
+            })
+            .ConfigureResource(rb => {
```
[the remark this answers is not in the document]
Comment (karolz-ms) [reply samples/eShopLite/ServiceDefaults/Extensions.cs:61]: I agree with the sentiment but honestly I am not sure if this can be done without some refactoring of OTLP configuration APIs. For now I just added it to the app templates.

**Comment by karolz-ms** [comment]
DCP ver 0.1.44 has been incorporated into the workload https://github.com/dotnet/aspire/commit/c0fac59fbb1fcbcf55a13878a539cda91a3dad6c This should be ready to merge now.

**Review by smitpatel — APPROVED** [review]

File: samples/eShopLite/ServiceDefaults/Extensions.cs [hunk 6]
```diff
@@ -56,6 +57,12 @@ public static IHostApplicationBuilder ConfigureOpenTelemetry(this IHostApplicati
...
+                rb.AddAttributes(new Dictionary<string, object>
+                {
+                    ["process.id"] = Environment.ProcessId,
```
Comment (davidfowl) [inline samples/eShopLite/ServiceDefaults/Extensions.cs:64]: cc @JamesNK

Comment (JamesNK) [reply samples/eShopLite/ServiceDefaults/Extensions.cs:64]: The OTEL way to identify resources is with `service.name`, `service.instance.id` and `service.namespace`. We should try to use them.

Spec: https://opentelemetry.io/docs/specs/semconv/resource/

File: src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs (hunk 2 above)
Comment (JamesNK) [reply src/Aspire.Hosting/Otlp/OtlpConfigurationExtensions.cs:41]: This change makes us follow the OTEL spec more correctly: https://opentelemetry.io/docs/specs/semconv/resource/#service

> MUST be the same for all instances of horizontally scaled services. If the value was not specified, SDKs MUST fallback to unknown_service: concatenated with [process.executable.name](https://opentelemetry.io/docs/specs/semconv/resource/process/#process), e.g. unknown_service:bash. If process.executable.name is not available, the value MUST be set to unknown_service.

File: samples/eShopLite/ServiceDefaults/Extensions.cs (hunk 6 above)
Comment (JamesNK) [reply samples/eShopLite/ServiceDefaults/Extensions.cs:64]: Some thoughts here:
* `service.instance.id` is a GUID that comes from the hosting UID. It's a great unique identifier from a computer's perspective, but lousy from a human's perspective.
* Resource attributes can be configured from env vars (it's what we are doing for `service.instance.id`) but that's not possible for PID because we don't know the `process.id` at this point.
* The problem with using PID as an identifier is hosting/dashboard can't define it ahead of time. See bullet point above.

Different approach:
* An alternative to OTEL sending the PID with its telemetry is matching the OTEL resource to the Aspire resource using the UID and getting the PID from the resource.
* How well is PID going to work in cloud scenarios?

File: src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs [hunk 7]
```diff
@@ -15,9 +16,11 @@ public class OtlpApplication
 {
     public const string SERVICE_NAME = "service.name";
     public const string SERVICE_INSTANCE_ID = "service.instance.id";
+    public const string PROCESS_ID = "process.pid";
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs:19]: Revert PID changes in OtlpApplication.

File: src/Aspire.Dashboard/Components/Pages/ConsoleLogs.razor.cs [hunk 8]
```diff
@@ -242,7 +242,29 @@ private static string GetDisplayText(ResourceViewModel resource)
         {
             stateText = $" ({resource.State})";
         }
-        return $"{resource.Name}{stateText}";
+        var resourceName = resource.DisplayName;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Pages/ConsoleLogs.razor.cs:245]: Console log and telemetry pages should display the same text in the dropdown. I think the console log dropdown should use the same format as telemetry.

File: src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        var byInstanceCount = applications.GroupBy(a => a.ApplicationName).ToDictionary(g => g.Key, g => g.Count());
+        var retval = applications.Select(a =>
+        {
+            var name = byInstanceCount[a.ApplicationName] > 1 ?
+                $"{a.ApplicationName} ({a.InstanceId.Substring(0, Math.Min(a.InstanceId.Length, 8))})"
+                : a.ApplicationName;
+            return new SelectViewModel<string> { Id = a.InstanceId, Name = name };
+        }).ToList();
+        return retval;
+    }
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs:21]: The same name should be in the grid, otherwise you have no idea what instance a log message was from:

image

There are many places that need to be updated. (logs grid, traces grid, tags, details view)

I'm ok with this being done in a follow up issue.

File: src/Aspire.Hosting.Dapr/DaprOptions.cs (hunk 1 above)
Comment (JamesNK) [reply src/Aspire.Hosting.Dapr/DaprOptions.cs:14]: Also, why is DaprOptions a record?

File: src/Aspire.Hosting.Dapr/CommandLineBuilder.cs [hunk 10]
```diff
@@ -130,6 +168,16 @@ private static CommandLineArgBuilder NamedStringArg(string name, string value, b
         };
     }
 
+    private static CommandLineArgBuilder ModelNamedStringArg(string name, string value, bool assignValue)
+    {
+        return () =>
+        {
+            return assignValue
+                ? new[] { $"{name}={value}" }
```
Comment (philliphoff) [inline src/Aspire.Hosting.Dapr/CommandLineBuilder.cs:176]: Are reserved characters not an issue in this case?  I'm not sure I follow the differences here.

File: src/Aspire.Hosting.Dapr/DaprOptions.cs (hunk 1 above)
Comment (philliphoff) [reply src/Aspire.Hosting.Dapr/DaprOptions.cs:14]: Where's the new `AddDapr()` interface?

File: src/Aspire.Dashboard/Components/Pages/ConsoleLogs.razor.cs (hunk 8 above)
Comment (karolz-ms) [reply src/Aspire.Dashboard/Components/Pages/ConsoleLogs.razor.cs:245]: If it means, use the OTel "service id" throughout the dashboard UI, I disagree. It is not meaningful to the user. The process ID works much better in this respect.

File: src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs (hunk 9 above)
Comment (karolz-ms) [reply src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs:21]: See my remark https://github.com/dotnet/aspire/pull/1184#discussion_r1415971387

File: src/Aspire.Hosting.Dapr/CommandLineBuilder.cs (hunk 10 above)
Comment (philliphoff) [reply src/Aspire.Hosting.Dapr/CommandLineBuilder.cs:176]: Ok, how about whitespace?

Comment (karolz-ms) [reply src/Aspire.Hosting.Dapr/CommandLineBuilder.cs:176]: Same thing. It is an API call; the values on both sides of the wire are/should be exactly the same.

File: src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs (hunk 7 above)
Comment (karolz-ms) [reply src/Aspire.Dashboard/Otlp/Model/OtlpApplication.cs]: Done

**Comment by joperezr** [comment]
Merged as[ CI was actually green](https://dev.azure.com/dnceng-public/public/_build/results?buildId=489526&view=logs&jobId=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a), there just seemed to be an issue with fetching back the status for some reason. We can report this if this happens again.

File: src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs (hunk 9 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs:21]: Please take another look at the screenshot. The resource column in the structured logs shows `catalogservice`. There is no indication which instance it is from. This needs to be resolved.

Comment (karolz-ms) [reply src/Aspire.Dashboard/Model/Otlp/SelectViewModelFactory.cs:21]: Makes a lot of sense. Would you like me to open an issue to track the work? I think we should take a look at all these places you mentioned in the dashboard UI and make it consistent between OTel-sourced data and other data.

**Comment by davidfowl** [comment]
Also looks the otel service name doesn't look right for the dapr scenario:

image

Somehow it's showing service B twice?
